src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -352,34 +352,33 @@ private IByteBuffer ReceiveBuffer(int responseTo, CancellationToken cancellation
             using (var receiveLockRequest = new SemaphoreSlimRequest(_receiveLock, cancellationToken))
             {
                 var messageTask = _dropbox.GetMessageAsync(responseTo);
-                try
+
+                Task.WaitAny(messageTask, receiveLockRequest.Task);
+                if (messageTask.IsCompleted)
                 {
-                    Task.WaitAny(messageTask, receiveLockRequest.Task);
-                    if (messageTask.IsCompleted)
-                    {
-                        return _dropbox.RemoveMessage(responseTo);
-                    }
+                    return _dropbox.RemoveMessage(responseTo);
+                }
 
-                    receiveLockRequest.Task.GetAwaiter().GetResult(); // propagate exceptions
-                    while (true)
+                receiveLockRequest.Task.GetAwaiter().GetResult(); // propagate exceptions
```

Comment:
I think in the current version, an exception, in this case, can be not handled correctly. I would suggest to safe the previous `try/catch` too. Just to avoid additional investigations.

Reply:
I have failed to reproduce the error in this line.
If something potentially fails here I would say that `messageTask.IsCompleted` would be `false` and enqueuing the buffer disposal would just lead into a hanging task and potentially another memory leak (I mean with the catch clause here: https://github.com/MikalaiMazurenka/mongo-csharp-driver/pull/58/files#diff-4325fc402c5272105bea0493a2bac7272debbf0c183046ae864f272128cd6d64L379).
So from my opinion the safest option is to leave the code as is and not revert the catch clause removal.
@rstam what do you think on this?

Reply:
Not sure what to think...

This line will almost never throw an exception. The only scenario I can think of where it might throw an exception is if cancellation is requested via the cancellationToken.

It would be worth figuring out what happens if the responseTo message then later arrives. Will it sit in the dropbox forever?

Reply:
I've implemented a test for this case https://github.com/DmitryLukyanov/mongo-csharp-driver/commit/6b3618fcc1dd82441dc820a30695a4cc16dd25e7#diff-4651f0118d6f84de1fea04c85e4ecf3675654b537defcca6ecec4a4a1af90057R413 

>The only scenario I can think of where it might throw an exception is if cancellation is requested via the cancellationToken.

yep, I think so too.

As far as I know, we cannot have such a situation in real life anymore. The only place when we use several receive methods one by one is a `streaming isMaster protocol`, but we use only single thread (task) for that operation. 

So, I think the current code is safe. Check my test and say what you think.

Reply:
Thanks for the tests Dmitry! They got me an idea of how it could be reproduced.
Let's assume we have two threads accessing the `BinaryConnection` simultaneously, **thread 1** which receives message 1 and **thread 2** which receives message 2.
Given thread 1 is the first one to acquires the lock [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L352):
```
var receiveLockRequest = new SemaphoreSlimRequest(_receiveLock, cancellationToken)
```
Then thread 2 attempts to acquire the lock second, calling he same line
```
var receiveLockRequest = new SemaphoreSlimRequest(_receiveLock, cancellationToken)
```
If cancellation is requested for thread 2 while it is inside `SemaphoreSlimRequest` constructor before the `SemaphoreSlim.WaitAsync` call [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Misc/SemaphoreSlimRequest.cs#L49) (meaning that `_linkedCancellationTokenSource.Token.IsCancellationRequested` is `true`):
```
_task = semaphore.WaitAsync(_linkedCancellationTokenSource.Token);
```
Then thread 2 `_task` will transit to `Cancelled` state right after the call to `WaitAsync` . Note: the cancellation token check inside `WaitAsync` is only done on the initial call to `WaitAsync`. The task wouldn't go to `Cancelled` state if cancellation is requested during the wait and in fact would transit to `RanToCompletion` once the lock is acquired).
Then for the thread 2 `Task.Wait` operation will be unblocked [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L357), continuing thread 2 execution:
```
Task.WaitAny(messageTask, receiveLockRequest.Task);
```
If the condition `messageTask.IsCompleted` is not satisfied [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L358) (meaning that thread 1 didn't read the message 2 yet):
```
if (messageTask.IsCompleted)
{
	return _dropbox.RemoveMessage(responseTo);
}
```
Then thread 2 will throw `OperationCanceledException` [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L363):
```
receiveLockRequest.Task.GetAwaiter().GetResult(); // propagate exceptions
```
After that if messages can arrive in arbitrary order then there's a chance that thread 1 will read the message 2 first and store it in the dropbox, then read message 2 and return it, thus leaving message 1 in the dropbox forever. This would happen if we don't have the catch clause [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L377) (and it is not present in current PR state).
But if we leave the catch clause in place, then thread 2 after throwing an exception [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L363):
```
receiveLockRequest.Task.GetAwaiter().GetResult(); // propagate exceptions
```
Will go to catch clause and schedule another thread to dispose a message once it arrives [here](https://github.com/mongodb/mongo-csharp-driver/blob/e82a1fb8bf7ce7f996f96517265a9cc0f7c4aa65/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L379):
```
var ignored = messageTask.ContinueWith(
	t => { _dropbox.RemoveMessage(responseTo).Dispose(); },
	TaskContinuationOptions.OnlyOnRanToCompletion);
```
And the statement `_dropbox.RemoveMessage(responseTo).Dispose();` will be executed as soon as thread 1 receives the message 2.